### ISSUE TITLE
[builds][Slow] openshift pipeline build Pipeline with maven slave [It] should build and complete successfully

### ISSUE THREAD

**Issue by coreydaley** [openshift/origin#13984]
[Fail] [builds][Slow] openshift pipeline build Pipeline with maven slave [It] should build and complete successfully 
/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/builds/pipeline.go:149

```
[INFO] Building war: /tmp/workspace/extended-test-jenkins-pipeline-0hvp5-dpbwt-openshift-jee-sample/target/ROOT.war
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.913 s
[INFO] Finished at: 2017-05-01T19:01:55+00:00
[INFO] Final Memory: 14M/202M
[INFO] ------------------------------------------------------------------------
[Pipeline] stash
Stashed 1 file(s)
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] node
Running on master in /var/lib/jenkins/jobs/extended-test-jenkins-pipeline-0hvp5-dpbwt-openshift-jee-sample/workspace
[Pipeline] {
[Pipeline] stage
[Pipeline] { (Build Image)
[Pipeline] unstash
[Pipeline] sh
[workspace] Running shell script
+ oc start-build openshift-jee-sample-docker --from-file=target/ROOT.war --follow -n extended-test-jenkins-pipeline-0hvp5-dpbwt
Uploading file "target/ROOT.war" as binary input for the build ...
Unable to connect to the server: net/http: TLS handshake timeout

```

https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended/285/consoleFull#-86936742456bf4006e4b05b79524e5923

I was unable to reproduce locally.

**Comment by stevekuznetsov** [comment]
Dupe of https://github.com/openshift/origin/issues/13983

**Comment by gabemontero** [comment]
OK, grepping on the build config `openshift-jee-sample-docker` in the master log, there are only gets on the bc, which I think call all be attributed to `config_controller.go`, and then the delete of the resource when the test ends and the namespace is cleaned up:

```
openshift.log:124420:I0501 14:59:53.300932    4209 config_controller.go:47] Handling BuildConfig extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:124421:I0501 14:59:53.300944    4209 config_controller.go:49] Handling Build Pruning: extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:124422:I0501 14:59:53.301549    4209 audit.go:133] 2017-05-01T14:59:53.30151033-04:00 AUDIT: id="d54ff94f-a541-497b-805d-23bbb6954c76" ip="172.18.0.216" method="GET" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:124424:I0501 14:59:53.302620    4209 wrap.go:75] GET /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (1.36226ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
openshift.log:125650:I0501 15:00:04.706976    4209 config_controller.go:47] Handling BuildConfig extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:125651:I0501 15:00:04.706985    4209 config_controller.go:49] Handling Build Pruning: extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:125652:I0501 15:00:04.707410    4209 audit.go:133] 2017-05-01T15:00:04.707385112-04:00 AUDIT: id="736923c1-92ce-4ed9-9ff9-99e39c6cf5bc" ip="172.18.0.216" method="GET" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:125654:I0501 15:00:04.708287    4209 wrap.go:75] GET /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (1.08399ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
openshift.log:130358:I0501 15:02:04.707679    4209 config_controller.go:47] Handling BuildConfig extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:130359:I0501 15:02:04.707688    4209 config_controller.go:49] Handling Build Pruning: extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:130360:I0501 15:02:04.708148    4209 audit.go:133] 2017-05-01T15:02:04.708097065-04:00 AUDIT: id="2470851f-0434-4a0d-b65c-190aaae61934" ip="172.18.0.216" method="GET" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:130362:I0501 15:02:04.709009    4209 wrap.go:75] GET /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (1.094511ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
openshift.log:135066:I0501 15:04:04.707574    4209 config_controller.go:47] Handling BuildConfig extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:135067:I0501 15:04:04.707586    4209 config_controller.go:49] Handling Build Pruning: extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:135068:I0501 15:04:04.708042    4209 audit.go:133] 2017-05-01T15:04:04.7080175-04:00 AUDIT: id="fef275b0-9866-4748-bb9d-f999a4bb382d" ip="172.18.0.216" method="GET" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:135070:I0501 15:04:04.708870    4209 wrap.go:75] GET /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (1.030516ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
openshift.log:138895:I0501 15:06:04.707571    4209 config_controller.go:47] Handling BuildConfig extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:138896:I0501 15:06:04.707579    4209 config_controller.go:49] Handling Build Pruning: extended-test-jenkins-pipeline-0hvp5-dpbwt/openshift-jee-sample-docker
openshift.log:138897:I0501 15:06:04.707985    4209 audit.go:133] 2017-05-01T15:06:04.70796178-04:00 AUDIT: id="bb43c1f0-f38b-4c34-8552-982a6d34601b" ip="172.18.0.216" method="GET" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:138899:I0501 15:06:04.708894    4209 wrap.go:75] GET /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (1.098852ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
openshift.log:140900:I0501 15:07:12.720613    4209 audit.go:133] 2017-05-01T15:07:12.720588278-04:00 AUDIT: id="14a8dcd1-8eb2-410a-b188-f82f849f6beb" ip="172.18.0.216" method="DELETE" user="system:openshift-master" groups="\"system:masters\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="extended-test-jenkins-pipeline-0hvp5-dpbwt" uri="/oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker"
openshift.log:140903:I0501 15:07:12.722806    4209 wrap.go:75] DELETE /oapi/v1/namespaces/extended-test-jenkins-pipeline-0hvp5-dpbwt/buildconfigs/openshift-jee-sample-docker: (2.37841ms) 200 [[openshift/v3.6.0 (linux/amd64) openshift/b48cd3c] 172.18.0.216:34570]
```

In other words, no POSTs that would have occured if the build config instantiatebinary endpoint had bubbled up to the master's `audit.go` or `wrap.go` level.

So things to pursue:

1) is there lower level networking type trace possible on either client or master side within the context of our extended tests running on jenkins, 
2) traversing through the k8s rest client code, I see a *very* sparse sprinkling of `glog.V(4)` trace; we could try adding a `--loglevel=10` to the `oc start-build` in the maven test, and see if that reveals anything

I'll start a PR for 2), and start asking around about 1)

**Comment by oatmealraisin** [comment]
Another failure [in the extended tests today](https://ci.openshift.redhat.com/jenkins/job/test_branch_origin_extended_builds/50/consoleFull#47332403656cbb9a5e4b02b88ae8c2f77).

An excerpt from the build logs shows some networking issues
```
Cloning repository https://github.com/openshift/jenkins-client-plugin.git
 > git init /var/lib/jenkins/jobs/extended-test-jenkins-pipeline-3spsj-q825m-sample-pipeline-openshift-client-plugin/workspace@script # timeout=10
Fetching upstream changes from https://github.com/openshift/jenkins-client-plugin.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/openshift/jenkins-client-plugin.git +refs/heads/*:refs/remotes/origin/*
 > git config remote.origin.url https://github.com/openshift/jenkins-client-plugin.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/openshift/jenkins-client-plugin.git # timeout=10
Fetching upstream changes from https://github.com/openshift/jenkins-client-plugin.git
 > git fetch --tags --progress https://github.com/openshift/jenkins-client-plugin.git +refs/heads/*:refs/remotes/origin/*
 > git rev-parse origin/master^{commit} # timeout=10
Checking out Revision ba6a931f77107189a796dc966425b08f8e710d5a (origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f ba6a931f77107189a796dc966425b08f8e710d5a
First time build. Skipping changelog.
[Pipeline] echo

[Pipeline] node
Running on master in /var/lib/jenkins/jobs/extended-test-jenkins-pipeline-3spsj-q825m-sample-pipeline-openshift-client-plugin/workspace
[Pipeline] {
[Pipeline] _OcContextInit
[Pipeline] readFile
[Pipeline] readFile
[Pipeline] _OcAction
[describe] 
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: Error during describe;
{reference={}, err=Unable to connect to the server: net/http: TLS handshake timeout, verb=describe, cmd=oc describe serviceaccount --server=https://172.30.0.1:443 --namespace=extended-test-jenkins-pipeline-3spsj-q825m --token=XXXXX , out=, status=1}

Finished: FAILURE
```Cloning repository https://github.com/openshift/jenkins-client-plugin.git
 > git init /var/lib/jenkins/jobs/extended-test-jenkins-pipeline-3spsj-q825m-sample-pipeline-openshift-client-plugin/workspace@script # timeout=10
Fetching upstream changes from https://github.com/openshift/jenkins-client-plugin.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/openshift/jenkins-client-plugin.git +refs/heads/*:refs/remotes/origin/*
 > git config remote.origin.url https://github.com/openshift/jenkins-client-plugin.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/openshift/jenkins-client-plugin.git # timeout=10
Fetching upstream changes from https://github.com/openshift/jenkins-client-plugin.git
 > git fetch --tags --progress https://github.com/openshift/jenkins-client-plugin.git +refs/heads/*:refs/remotes/origin/*
 > git rev-parse origin/master^{commit} # timeout=10
Checking out Revision ba6a931f77107189a796dc966425b08f8e710d5a (origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f ba6a931f77107189a796dc966425b08f8e710d5a
First time build. Skipping changelog.
[Pipeline] echo

[Pipeline] node
Running on master in /var/lib/jenkins/jobs/extended-test-jenkins-pipeline-3spsj-q825m-sample-pipeline-openshift-client-plugin/workspace
[Pipeline] {
[Pipeline] _OcContextInit
[Pipeline] readFile
[Pipeline] readFile
[Pipeline] _OcAction
[describe] 
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: Error during describe;
{reference={}, err=Unable to connect to the server: net/http: TLS handshake timeout, verb=describe, cmd=oc describe serviceaccount --server=https://172.30.0.1:443 --namespace=extended-test-jenkins-pipeline-3spsj-q825m --token=XXXXX , out=, status=1}

Finished: FAILURE
```

The jenkins logs also has repeated connection exceptions thrown,

```
May 03, 2017 1:37:30 PM io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager nextReconnectInterval
INFO: Current reconnect backoff is 1000 milliseconds (T0)
May 03, 2017 1:37:30 PM io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager closeWebSocket
WARNING: Failed to write websocket close message: class java.net.SocketException Socket is closed
May 03, 2017 1:37:30 PM io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$2 onFailure
WARNING: Exec Failure
java.net.SocketException: Connection closed by remote host
	at sun.security.ssl.SSLSocketImpl.checkWrite(SSLSocketImpl.java:1565)
	at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:71)
	at okio.Okio$1.write(Okio.java:78)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
	at okio.RealBufferedSink.emit(RealBufferedSink.java:178)
	at okhttp3.internal.ws.WebSocketWriter.writeControlFrameSynchronized(WebSocketWriter.java:152)
	at okhttp3.internal.ws.WebSocketWriter.writePing(WebSocketWriter.java:79)
	at okhttp3.internal.ws.RealWebSocket.sendPing(RealWebSocket.java:143)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$2$1.execute(WatchConnectionManager.java:171)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$NamedRunnable.run(WatchConnectionManager.java:424)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by gabemontero** [comment]
In chatting with @oatmealraisin and digesting the data he noted, of particular interest to me was:

```
ERROR: Error during describe;
{reference={}, err=Unable to connect to the server: net/http: TLS handshake timeout, verb=describe, cmd=oc describe serviceaccount --server=https://172.30.0.1:443 --namespace=extended-test-jenkins-pipeline-3spsj-q825m --token=XXXXX , out=, status=1}
```

Is exactly like the error seen previously with `oc start-build`, only with `oc describe` this time.  

I agree the java side exceptions @oatmealraisin noted seem similar.

We'll have to consider when diving into this more whether this issue is somehow Jenkins/JVM specific, or if the Jenkins/JVM scenario is surfacing a new instability in openshift/origin.

**Comment by gabemontero** [comment]
Perhaps a coincidence and something else entirely (since a tls error is pretty low level), but I just saw this tls error running a local unit test:

```
gmontero ~/go/src/github.com/openshift/origin  (pipeline-logs-client-err-2-url)$ hack/test-go.sh pkg/cmd/cli/cmd
--- FAIL: TestSetPipelineBuild (0.00s)
	logs_test.go:111: &{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] } {{ <nil> <nil> <nil> {map[] map[]} map[] map[] <nil>} 0 [] 0 <nil> false false map[] <nil>} {0 0 0 0 0 0 <nil> [] 0}}, unexpected results isPipeline false JenkinsLogs true
2017-05-03 13:02:29.801786 I | http: TLS handshake error from 127.0.0.1:48424: EOF
2017-05-03 13:02:29.819717 I | http: TLS handshake error from 127.0.0.1:49948: EOF
Tag yourproject/rails:tip set to openshift/ruby:latest.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Deleted tag yourproject/rails:tip.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Deleted tag yourproject/rails:tip.
FAIL
coverage: 19.6% of statements
FAIL	github.com/openshift/origin/pkg/cmd/cli/cmd	4.954s
hack/test-go.sh failed after 15 seconds
```

**Comment by gabemontero** [comment]
Possible correlation ... in debugging some new unit tests, if an api object was invalid, rather than getting a validation error, I got a tls handshake error:

```
--- FAIL: TestSetPipelineBuild (0.00s)
	logs_test.go:111: &{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] } {{ <nil> <nil> <nil> {map[] map[]} map[] map[] <nil>} 0 [] 0 <nil> false false map[] <nil>} {0 0 0 0 0 0 <nil> [] 0}}, unexpected results isPipeline false JenkinsLogs true
2017-05-03 13:02:29.801786 I | http: TLS handshake error from 127.0.0.1:48424: EOF
2017-05-03 13:02:29.819717 I | http: TLS handshake error from 127.0.0.1:49948: EOF
Tag yourproject/rails:tip set to openshift/ruby:latest.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Deleted tag yourproject/rails:tip.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Tag yourproject/rails:tip set to openshift/ruby:2.0.
Deleted tag yourproject/rails:tip.
FAIL
```

I wonder if at least for the failed `oc` calls, if some other error is getting masked as tls handshake ....

**Comment by gabemontero** [comment]
I am seeing another flavor of this test failing.  Specifically, it looks like the `oc start-build --follow` is not exiting because of the `--follow`.  I've seen this when running `oc start-build openshift-jee-sample-docker --follow` manually.  The build finished but the `oc ... --follow` does not exit and I have to control-c.

I've also reproduced this locally/manually in jenkins deployed in openshift.  It is not everytime, but has happened more than once.

Also, on the point of "what might have changed" ... we did update the oc client in the jenkins image late last week:  see https://github.com/openshift/jenkins/commit/d48c23c5461dacf95042c883a391d02fee834a2a

I wonder if that is at least part of the instability we've seen, at least in the cases where invocations of `oc` are involved (which includes our extended tests and the openshift-client plugin).  It would exclude sync plugin / fabric8 ssl comm issues, and the master seeing readiness check failures for the jenkins pod.

@bparees wdyt ^^

**Comment by gabemontero** [comment]
Also note, `oc logs` really defers to k8s logs for the actual streaming, etc.  It potentially could have been impacted by the rebase.

**Comment by stevekuznetsov** [comment]
@ncdc could probably weigh in on `kubectl logs` and/or `spdystream`

**Comment by ncdc** [comment]
We don't use spdy for retrieving logs. Only exec, attach, and portforward.

There is a kube PR to fix an issue where --follow wouldn't stop when the container exited, but as best I can tell, it only applies to the json-file docker log driver. https://github.com/kubernetes/kubernetes/pull/44406. 

If you start a pod that does something like print out the date every second for five or ten seconds before terminating, and then in another terminal run `oc logs mypod -f`, does it stop following, or does it hang waiting for more logs?

**Comment by gabemontero** [comment]
On Thu, May 4, 2017 at 8:19 PM Andy Goldstein <notifications@github.com>
wrote:

> We don't use spdy for retrieving logs. Only exec, attach, and portforward.
>
> There is a kube PR to fix an issue where --follow wouldn't stop when the
> container exited, but as best I can tell, it only applies to the json-file
> docker log driver. kubernetes/kubernetes#44406
> <https://github.com/kubernetes/kubernetes/pull/44406>.
>
> If you start a pod that does something like print out the date every
> second for five or ten seconds before terminating, and then in another
> terminal run oc logs mypod -f, does it stop following, or does it hang
> waiting for more logs?
>
The frequency of logging was similar to what you described, and it hangs
waiting for more logs.  In other words The pod was done generating output
(for an openshift build in our case), and all the logs showed where the
--follow was employed.

The build pod / container had exited as well.

**Comment by ncdc** [comment]
Would you have time to try to reproduce with the smallest example possible,
like what I described above with just a simple pod?

**Comment by gabemontero** [comment]
Separate from the --follow item discussed here, on the debug front For the TLS issues I'm planning to 

- ramp the extended test scope ...  the focused scope has been too successful 
- on the jenkins image commit I noted earlier, in addition to a bump in oc version there was a bump in jenkins version; assuming I can get at least intermittent failures I am the going to run with custom jenkins images that revert first jenkins, then oc, and see if that has bearing

**Comment by gabemontero** [comment]
On Thu, May 4, 2017 at 8:36 PM Andy Goldstein <notifications@github.com>
wrote:

> Would you have time to try to reproduce with the smallest example possible,
> like what I described above with just a simple pod?


Sure.  I've got some diagnostic ideas I just posted for the more prevalent
TLS comm errors, but when I get done with those I'll circle back and try a
simple pod.

**Comment by gabemontero** [comment]
@ncdc - here is what I believe is a simple pod example:  
[pod.json.zip](https://github.com/openshift/origin/files/979643/pod.json.zip)

Details on my locally build origin:

```
gmontero ~/go/src/github.com/openshift/origin  (maven-ext-test-dbg)$ openshift version
openshift v3.6.0-alpha.1+999c737-400
kubernetes v1.6.1+5115d708d7
etcd 3.1.0
gmontero ~/go/src/github.com/openshift/origin  (maven-ext-test-dbg)$ oc version
oc v3.6.0-alpha.1+999c737-400
kubernetes v1.6.1+5115d708d7
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://127.0.0.1:8443
openshift v3.6.0-alpha.1+999c737-400
kubernetes v1.6.1+5115d708d7
gmontero ~/go/src/github.com/openshift/origin  (maven-ext-test-dbg)$ 
```

This is at some level I built yesterday (with a debug commit on top).

This particular pod ends up error'ing out after running for a bit, but since I've seen this occur repeatedly with both pods that complete after a while and pods that error out after a while, I figured it wasn't worth debugging this sample further. 

I'm going to rebase / rebuild my origin in a sec from latest master, to see if the error still occurs.  I'll report the finding when I have them.

thanks

**Comment by ncdc** [comment]
@gabemontero I haven't tested with a build pod, but I have tested with a simple centos:centos7, and the `oc logs -f` command terminates as soon as the container exits.

**Comment by gabemontero** [comment]
On Fri, May 5, 2017 at 1:22 PM Andy Goldstein <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> I haven't tested with a
> build pod, but I have tested with a simple centos:centos7, and the oc
> logs -f command terminates as soon as the container exits.
>
@ncdc yep ther is an intermittent nature to this.  I've seen it
consistently on my fedora 25 laptop but only occasionally in our extended
tests on ci.openshift.  Not sure how to pin down the environmental
differences.

**Comment by gabemontero** [comment]
On Fri, May 5, 2017 at 1:30 PM Gabe Montero <gmontero@redhat.com> wrote:

> On Fri, May 5, 2017 at 1:22 PM Andy Goldstein <notifications@github.com>
> wrote:
>
>> @gabemontero <https://github.com/gabemontero> I haven't tested with a
>> build pod, but I have tested with a simple centos:centos7, and the oc
>> logs -f command terminates as soon as the container exits.
>>
> @ncdc yep ther is an intermittent nature to this.  I've seen it
> consistently on my fedora 25 laptop but only occasionally in our extended
> tests on ci.openshift.  Not sure how to pin down the environmental
> differences.
>

Perhaps I should see if I can induce a golang thread dump on the oc logs on
my system ....would that be of any use?

**Comment by ncdc** [comment]
Definitely. `kill -3`. You may need to redirect the output of `oc logs` to
a file in case the thread dump overflows your terminal scrollback.

**Comment by gabemontero** [comment]
@ncdc here is the thread dump:  
[thread-dump.txt](https://github.com/openshift/origin/files/980063/thread-dump.txt)

Other than goroutine 72, everything else is blocked on IO or semaphores down in golang code.

The comment before the channel select in the apimachinery code was ominous at first glance:
```
		// NOTE: b/c there is no priority selection in golang
		// it is possible for this to race, meaning we could
		// trigger t.C and stopCh, and t.C select falls through.
		// In order to mitigate we re-check stopCh at the beginning
		// of every loop to prevent extra executions of f().
```
Though that logic has not changed in a while if I'm reading `git blame` correctly.

Here is my go version:  go version go1.7.3 linux/amd64

first blush, smells like a golang issue

(admittedly novice) opinion, I would contend some defensive programming is in order, andthat the call to `io.Copy` should be coupled with a select/channel/timer so we don't block forever ... thoughts?

**Comment by gabemontero** [comment]
Older PR test and overnight failures have been pruned, but between the debug runs in https://github.com/openshift/origin/pull/14013, a recent overnight run that resulted in https://github.com/openshift/origin/issues/14093, 
and @coreydaley 's failure noted in https://github.com/openshift/origin/issues/14084, I'm not seeing the common thread of volume mount issues.  In the case of the client plugin, they result in the mongodb deployment failing.  Here is a snippet from @coreydaley 's run cataloged in https://github.com/openshift/origin/issues/14084:

```
May  5 15:48:46.203: INFO: At 2017-05-05 15:42:23.452467294 -0400 EDT - event for jenkins-1-deploy: {kubelet 172.18.0.186} FailedMount: Unable to mount volumes for pod "jenkins-1-deploy_extended-test-jenkins-pipeline-n2w9s-r8pf3(99bb5368-31ca-11e7-ab80-0e76db0f9e3c)": timeout expired waiting for volumes to attach/mount for pod "extended-test-jenkins-pipeline-n2w9s-r8pf3"/"jenkins-1-deploy". list of unattached/unmounted volumes=[deployer-token-0qmcm]
May  5 15:48:46.203: INFO: At 2017-05-05 15:42:23.452535031 -0400 EDT - event for jenkins-1-deploy: {kubelet 172.18.0.186} FailedSync: Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "extended-test-jenkins-pipeline-n2w9s-r8pf3"/"jenkins-1-deploy". list of unattached/unmounted volumes=[deployer-token-0qmcm]
May  5 15:48:46.203: INFO: At 2017-05-05 15:42:30.954662138 -0400 EDT - event for mongodb-1-77ttj: {kubelet 172.18.0.186} Started: Started container with id 428c475c64b8b41904775b28744ca7793090960feb08376ccfd54dd5f8a60677
May  5 15:48:46.203: INFO: At 2017-05-05 15:42:53.009580387 -0400 EDT - event for mongodb-1-deploy: {kubelet 172.18.0.186} Killing: Killing container with id docker://95b68fb656cf1c0c17e542ab73123629e84885caae3cc6133bcc673864b6b479:Need to kill Pod
May  5 15:48:46.203: INFO: At 2017-05-05 15:44:52.985650286 -0400 EDT - event for mongodb-1-deploy: {kubelet 172.18.0.186} FailedMount: Unable to mount volumes for pod "mongodb-1-deploy_extended-test-jenkins-pipeline-n2w9s-r8pf3(c7fe9baa-31ca-11e7-ab80-0e76db0f9e3c)": timeout expired waiting for volumes to attach/mount for pod "extended-test-jenkins-pipeline-n2w9s-r8pf3"/"mongodb-1-deploy". list of unattached/unmounted volumes=[deployer-token-0qmcm]
May  5 15:48:46.203: INFO: At 2017-05-05 15:44:52.985704605 -0400 EDT - event for mongodb-1-deploy: {kubelet 172.18.0.186} FailedSync: Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "extended-test-jenkins-pipeline-n2w9s-r8pf3"/"mongodb-1-deploy". list of unattached/unmounted volumes=[deployer-token-0qmcm]
```

I'm not ready to close this issue out in lieu of https://github.com/openshift/origin/issues/14093 until

a) we get a good week or so of runs where a failure occurs and this apparent mount issues is not in play ... though admittedly, it is entirely conceivable that these were present in the older runs and I just missed them / wasn't cognizant enough earlier on to look for them

b) the `oc logs` | `oc start-build --follow` not returning phenomenon is still in theory pending ... minimally, as a fall back I'm thinking we switch the `oc start-build --follow` to `oc start-build --wait` in our jee jenkins pipeline sample

**Comment by gabemontero** [comment]
FYI - broke out the log following thread here to separate issue https://github.com/openshift/origin/issues/14148

**Comment by gabemontero** [comment]
the latest findings in https://github.com/openshift/origin/pull/14013 are pointing toward a solution for this issue

if the trend in the PR continues for another day or so, I'll rebrand it as a fix
